### DISCUSSION TITLE
Ethernet on NEORV32 SoC

### DISCUSSION THREAD

**Discussion by hungdinhthe** [stnolting/neorv32#401] [Q&A]
Hi, I want to build a web server on NEORV32 SoC with freertos. I am trying to undestand how to connect NEORV32 SoC to ethernet via WISHBONE (Axi 4 Lite). My board has an RMII connection for Ethernet board so I am looking at Xilinx Axi4 Lite Ethernet IP (It work as an ethernet MAC). I was wondering if I connect the Ethernet Mac to the NEORV32 via WISHBONE, how does it recognize the ethernet mac and do I need additional driver for TCP/IP connection.
An example or basic instruction to understand further into this matter will help me a lot.
Thank you.

**Comment by stnolting** [comment]
Hey there!

That sounds like an interesting project - I always wanted to build something like that, too!

I would start like this:
* first, implement the "glue logic" to connect the Xilinx Ethernet IP to the processor; most IP modules can be configured to provide an AXI-Lite interface, which should connect without problems to the NEORV32 AXI-Lite wrapper
* then, I would try to implement the low-level accesses on the software side: so, reading and writing register in the Ethernet core
* if the low level stuff works, you could check out some example setups (there are a lot of Microblaze examples implementing a web server) and try to port the "drivers"
* there are a lot of user guides and application notes (like https://support.xilinx.com/s/article/38280?language=en_US) that should help implementing low-level accesses

I am not too familiar with FreeRTOS and Ethernet. But once you have the low-level drivers running, integration into FreeRTOS shouldn't be too complicated.
